File: carode.py
import numpy as np
import scipy.integrate
import pandas as pd
import altair as alt

clamp1 = lambda x: min(1, max(x, -1))

tmax = 20.0
tmax = 4.5e-11

g = 9.81  # [m/s^2] gravitational acceleration
wheel_radius = 0.34  # [m]
inertia = 5  # [kg m] moment of inertia of wheel (+drive line)
mass = 1500  # [kg] mass of the care


def sigma(omega, v):
    """Compute the slip ratio"""
    if v != 0:
        return (omega * wheel_radius - v) / abs(v)
    elif omega == 0:
        return 0
    else:
        return np.sign(omega) * np.inf


def t_drive(t):
    """Torque from the engine onto the wheel, in Nm"""
    if t < 3:
        return 200
    elif t < 4:
        return 0
    elif t < 7:
        return -500
    elif t < 8:
        return 5500
    elif t < 13:
        return 2000
    else:
        return 0


def f_drag(v):
    C = 2  # drag constant dependant on geometry, surface etc
    return -C * v * np.abs(v)


def f_roll(v):
    C = 0.7  # rolling resistance constant depends on stuffs
    return -C * v


def func(t, y, t_drive):
    v, omega = y
    slip_ratio = sigma(omega, v)
    f_traction = clamp1(slip_ratio / 0.06) * mass * g
    a = (f_traction + f_drag(v) + f_roll(v)) / mass
    alpha = (-f_traction * wheel_radius + t_drive(t)) / inertia
    return [a, alpha]


# LSODA can switch between stiff and nonstiff solver
# to get precision AND speed
method = "LSODA"
y0 = np.array([0, 0])
args = (t_drive,)
res = scipy.integrate.solve_ivp(
    fun=func, y0=y0, t_span=[0, tmax], args=args, method=method
)
y = res.y
t = res.t
print(res.message)
print(f"Solving took {res.nfev} function evaluations to complete")
v = y[0, :]
omega = y[1, :]
[as_,alphas] = list(zip(*[func(t,y,t_drive) for t,y in zip(t,y.T)]))


#
# Plot the results!
#
C_t = "Time (s)"
C_v = "Car speed (m/s)"
C_omegaR = "Wheel speed (m/s)"
C_drive = "Drive torque (kNm)"
C_roll = "Roll resistance (kN)"
C_drag = "Drag force (kN)"
C_sr = "Slip ratio (%)"
C_acc = "Acceleration (m/s^2)"
C_wAcc = "Wheel Acceleration (m/s^2)"
data = pd.DataFrame(
    {
        C_t: t,
        C_v: v,
        C_omegaR: omega * wheel_radius,
        C_drive: [t_drive(s) / 1000 for s in t],
        C_drag: [f_drag(w) / 1000 for w in v],
        C_roll: [f_roll(w) / 1000 for w in v],
        C_sr: [100 * sigma(om, ve) for om, ve in zip(omega, v)],
        C_acc: as_,
        C_wAcc: [alpha*wheel_radius for alpha in alphas],
    }
)
alt.Chart(data).transform_fold(
    fold=[C_v, C_omegaR, C_drive, C_sr, C_drag, C_roll, C_acc, C_wAcc], as_=["Data", "Value"]
).mark_line(point=alt.OverlayMarkDef()).encode(
    alt.Y("Value", type="quantitative", scale=alt.Scale(domain=(-2, 10))),
    alt.X(C_t),
    alt.Color("Data", type="nominal"),
).properties(
    width=600, height=600
).interactive().show()

File: examples/CarOde.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE TypeFamilies #-}

import Control.Monad
import Data.List
import GHC.Real
import Graphics.Vega.VegaLite
import LSODA
    ( simpLsoda,
      LSODARes(LSODARes, success, ts, ys, msg, optOutput),
      OptOut(LSODAOO, nfe),
      RHS(..),
      TimeSpec(StartStop) )
import Text.Printf
import Numeric.LinearAlgebra.Static
import qualified Numeric.LinearAlgebra.Data as LAD



g :: Double
g = 9.81 --[m/s^2] gravitational acceleration

wheelRadius :: Double
wheelRadius = 0.34 --[m]

inertia :: Double
inertia = 5 --[kg m] moment of inertia of wheel (+drive line)

mass :: Double
mass = 1500 --[kg] mass of the care

tMax :: Double
tMax = 20

inf :: Double
inf = fromRational infinity

clamp1 :: Double -> Double
clamp1 x = min 1 $ max x (-1)

sigma :: Double -> Double -> Double
sigma 0 0 = 0
sigma omega 0
  | omega < 0 = -inf
  | otherwise = inf
sigma omega v = (omega * wheelRadius - v) / abs v

tDrive :: Double -> Double
tDrive t
  | t < 3 = 200
  | t < 4 = 0
  | t < 7 = -500
  | t < 8 = 5500
  | t < 13 = 2000
  | otherwise = 0

fDrag :: Double -> Double
fDrag v = (-c) * v * abs v
  where
    c = 2 -- air resistence constant

fRoll :: Double -> Double
fRoll v = (-c) * v
  where
    c = 0.7 -- rolling resistance constant

fprim :: RHS 2
fprim = MkRHS fprim'

fprim' :: (Double -> R 2 ->  R 2)
fprim' t y =
  let (v, tmp) = headTail y
      (omega,_) = headTail tmp
      slipRatio = sigma omega v
      fTraction = clamp1 (slipRatio / 0.06) * mass * g
      a = (fTraction + fDrag v + fRoll v) / mass
      alpha = (-fTraction * wheelRadius + tDrive t) / inertia
  in vector [a, alpha]

main :: IO ()
main = do
  let c_t = "Time (s)"
  let c_v = "Car speed (m/s)"
  let c_acc = "Acceleration (m/s)"
  let c_wAcc = "Wheel acceleration (m/s^2)"
  let c_omegaR = "Wheel speed (m/s)"
  let c_drive = "Drive torque (kNm)"
  let c_roll = "Roll resistance (kN)"
  let c_drag = "Drag force (kN)"
  let c_sr = "Slip ratio (%)"
  let res = simpLsoda fprim (0 :: R 2 ) (StartStop 0 tMax)
  let LSODARes {success = didSucceed, ts = t, ys = ys', msg = msg', optOutput = LSODAOO {nfe = feval}} = res
  unless didSucceed . printf "Failed solving. Message: %s" $ msg'
  printf "It took %d function evaluations to complete all" feval
  let [v, omega] = transpose $ map (LAD.toList . unwrap) ys'
  let ydots = zipWith (\ a b -> LAD.toList ( unwrap ( (unRHS fprim) a b))) t ys'
  let [a, alpha] = transpose ydots
  let omegaR = [w * wheelRadius | w <- omega]
  let wAcc = [alpha' * wheelRadius | alpha' <- alpha]
  let manualData =
        dataFromColumns []
          . dataColumn c_t (Numbers t)
          . dataColumn c_v (Numbers v)
          . dataColumn c_acc (Numbers a)
          . dataColumn c_wAcc (Numbers wAcc)
          . dataColumn c_omegaR (Numbers omegaR)
          . dataColumn c_drive (Numbers [tDrive s / 1000 | s <- t])
          . dataColumn c_drag (Numbers (map (\v' -> fDrag v' / 1000) v))
          . dataColumn c_roll (Numbers (map (\w' -> fRoll w' / 1000) omega))
          . dataColumn c_sr (Numbers (map (100 *) (zipWith sigma omega v)))
          $ []
  let trans =
        transform
          . foldAs
            [ c_v,
              c_omegaR,
              c_drive,
              c_sr,
              c_drag,
              c_roll,
              c_acc,
              c_wAcc
            ]
            "Data"
            "Value"
  let enc =
        encoding
          . position X [PName "Time (s)", PmType Quantitative]
          . position Y [PName "Value", PmType Quantitative, PScale [SDomain (DNumbers [-2, 10])]]
          . color [MName "Data", MmType Nominal]
  let sel = selection . select "view" Interval [BindScales] $ []
  let vlPlot =
        toVegaLite
          [ manualData,
            trans [],
            mark Line [MPoint (PMMarker [])],
            enc [],
            height 600,
            width 600,
            sel
          ]
  toHtmlFile "plot.html" vlPlot
  putStrLn "Wrote plot.html. Check it!"